### PR TITLE
Dragon tags & mechs

### PR DESCRIPTION
**Pull request by heypr**
- Requested by [Josh65](https://discord.com/channels/315163488085475337/1231813740538630217).
- Adds 3 paper-only, 1.20+ tags:
  - `gateway_count`
  - `healing_crystals` 
  - `respawn_crystals`
- Adds 1 paper-only, 1.20+ mechanism:
  - `spawn_gateway`
- Adds 1 regular tag:
  - `first_dragon_killed`
- Adds 4 regular mechanisms;
  - `respawn_dragon`
  - `reset_crystals`
  - `respawn_phase`
  - `first_dragon_killed`

I opened this as a draft PR because I haven't tested any of this/just wrote this up, and want it open so that I don't lose it in my sea of local and published branches. Feel free to comment/review though!

### CODE REVIEW

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 1]
```diff
@@ -23,6 +29,76 @@ public static void register() {
             return new ElementTag(world.getWorld().getNoTickViewDistance());
         });
 
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_20)) {
+            // <--[tag]
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java:34]: should have newline above this meta start

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 2]
```diff
@@ -23,6 +29,76 @@ public static void register() {
...
+            // @Plugin Paper
+            // @group paper
+            // @description
+            // If no location is specified, tries to spawn a new end gateway using default game mechanics. Otherwise, spawns a new end gateway portal at the specified location.
+            // Only works in end worlds.
+            // -->
+            WorldTag.tagProcessor.registerMechanism("spawn_gateway", false, (object, mechanism) -> {
+                DragonBattle battle = object.getWorld().getEnderDragonBattle();
+                if (battle == null) {
+                    return;
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java:103]: should have an echoError

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 3]
```diff
@@ -23,6 +29,76 @@ public static void register() {
...
+                }
+                if (!mechanism.hasValue()) {
+                    battle.spawnNewGateway();
+                }
+                if (mechanism.requireObject(LocationTag.class)) {
```
Comment (mcmonkey4eva) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java]: `else if`

File: plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java [hunk 4]
```diff
@@ -1006,6 +1007,103 @@ else if (time >= 12500) {
...
+        // <--[mechanism]
+        // @object WorldTag
+        // @name respawn_dragon
+        // @description
+        // Initiates the respawn sequence of the ender dragon as if a player placed 4 end crystals on the portal.
+        // Only works in end worlds.
+        // -->
+        tagProcessor.registerMechanism("respawn_dragon", false, (object, mechanism) -> {
+            if (object.getWorld() == null) {
+                mechanism.echoError("World '" + object.world_name + "' is unloaded, cannot set respawn phase.");
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java]: ... when is this a real concern?

File: plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java [hunk 5]
```diff
@@ -1006,6 +1007,103 @@ else if (time >= 12500) {
...
+            if (battle == null) {
+                return;
+            }
+            battle.resetCrystals();
+        });
+
+        // <--[mechanism]
+        // @object WorldTag
+        // @name respawn_phase
+        // @input ElementTag(String)
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java]: just `ElementTag`

File: plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java [hunk 6]
```diff
@@ -1006,6 +1007,103 @@ else if (time >= 12500) {
...
+        // Only works in end worlds.
+        // -->
+        tagProcessor.registerMechanism("reset_crystals", false, (object, mechanism) -> {
+            if (object.getWorld() == null) {
+                mechanism.echoError("World '" + object.world_name + "' is unloaded, cannot reset crystals.");
+                return;
+            }
+            DragonBattle battle = object.getWorld().getEnderDragonBattle();
+            if (battle == null) {
+                return;
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java:1056]: these ones need an echoError too

**Review by mcmonkey4eva — DISMISSED** [review]

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 7]
```diff
@@ -23,6 +29,81 @@ public static void register() {
...
         });
 
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_20)) {
+
+            // <--[tag]
+            // @attribute <WorldTag.gateway_count>
+            // @returns ElementTag(Number)
+            // @description
+            // Returns the number of end gateway portals.
+            // Only works in end worlds.
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java]: Minor nitpick, but `in` sounds more like the player being in an end world while the script is running or something - maybe `on`, `for`, `if the world is an end world`, etc.?

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 8]
```diff
@@ -23,6 +29,81 @@ public static void register() {
...
+                    return null;
+                }
+                return new ElementTag(battle.getGatewayCount());
+            });
+
+            // <--[tag]
+            // @attribute <WorldTag.healing_crystals>
+            // @returns ListTag(EntityTag)
+            // @description
+            // Returns a ListTag of the healing crystals located on top of the obsidian towers.
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java]: I'd mention ender/end crystals as a search aid or in the description somewhere, healing crystals is technically right but most people probably won't know them by that name.

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 9]
```diff
@@ -23,6 +29,81 @@ public static void register() {
...
+                DragonBattle battle = object.getWorld().getEnderDragonBattle();
+                if (battle == null) {
+                    attribute.echoError("Provided world is not an end world!");
+                    return null;
+                }
+                return new ListTag(battle.getHealingCrystals(), EntityTag::new);
+            });
+
+            // <--[tag]
+            // @attribute <WorldTag.healing_crystals>
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java]: Wrong tag name

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 10]
```diff
@@ -23,6 +29,81 @@ public static void register() {
...
+            });
+
+            // <--[mechanism]
+            // @object WorldTag
+            // @name spawn_gateway
+            // @input LocationTag
+            // @Plugin Paper
+            // @group paper
+            // @description
+            // If no location is specified, tries to spawn a new end gateway using default game mechanics. Otherwise, spawns a new end gateway portal at the specified location.
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java]: Nitpick, but meta is usually the other way around (`spawns gateway at specified location, specify no input to spawn vanilla one`)

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 11]
```diff
@@ -23,6 +29,81 @@ public static void register() {
...
+                DragonBattle battle = object.getWorld().getEnderDragonBattle();
+                if (battle == null) {
+                    mechanism.echoError("Cannot spawn gateway in non-end world!");
+                    return;
+                }
+                if (!mechanism.hasValue()) {
+                    battle.spawnNewGateway();
+                }
+                else if (mechanism.requireObject(LocationTag.class)) {
+                    battle.spawnNewGateway(mechanism.getValue().asType(LocationTag.class, mechanism.context));
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java]: There should be a `#valueAsType` iirc?

File: plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java [hunk 12]
```diff
@@ -1006,6 +1008,92 @@ else if (time >= 12500) {
...
+            }
+            battle.resetCrystals();
+        });
+
+        // <--[mechanism]
+        // @object WorldTag
+        // @name respawn_phase
+        // @input ElementTag
+        // @description
+        // Set the current respawn phase of the ender dragon. Valid phases can be found at <@link url https://jd.papermc.io/paper/1.20/org/bukkit/boss/DragonBattle.RespawnPhase.html>
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java]: Link doesn't seem to work?

File: plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java [hunk 13]
```diff
@@ -1006,6 +1008,92 @@ else if (time >= 12500) {
...
+        // Only works in end worlds.
+        // -->
+        tagProcessor.registerMechanism("respawn_phase", false, ElementTag.class, (object, mechanism, input) -> {
+            DragonBattle battle = object.getWorld().getEnderDragonBattle();
+            if (battle == null) {
+                mechanism.echoError("Provided world is not an end world!");
+                return;
+            }
+            battle.setRespawnPhase(input.asEnum(DragonBattle.RespawnPhase.class));
+        });
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java]: Needs `Mechanism#requireEnum`

File: plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java [hunk 14]
```diff
@@ -1006,6 +1008,92 @@ else if (time >= 12500) {
...
+
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_20)) {
+
+            // <--[mechanism]
+            // @object WorldTag
+            // @name respawn_phase
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java]: Wrong name

File: plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java [hunk 15]
```diff
@@ -1006,6 +1008,92 @@ else if (time >= 12500) {
...
+            // Only works in end worlds.
+            // -->
+            tagProcessor.registerMechanism("first_dragon_killed", false, ElementTag.class, (object, mechanism, input) -> {
+                DragonBattle battle = object.getWorld().getEnderDragonBattle();
+                if (battle == null) {
+                    mechanism.echoError("Provided world is not an end world!");
+                    return;
+                }
+                battle.setPreviouslyKilled(input.asBoolean());
+            });
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java]: Needs `Mechanism#requireBoolean`

File: plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java [hunk 16]
```diff
@@ -1006,6 +1008,92 @@ else if (time >= 12500) {
...
+
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_20)) {
+
+            // <--[mechanism]
+            // @object WorldTag
+            // @name respawn_phase
+            // @input ElementTag(Boolean)
+            // @description
+            // Set whether the first ender dragon was killed already.
+            // Only works in end worlds.
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java]: Might be nice to document exactly what this does - does it reset the end and respawn the dragon? is it just a value that doesn't really affect anything?

**Review by mcmonkey4eva — DISMISSED** [review]

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 17]
```diff
@@ -23,6 +29,82 @@ public static void register() {
...
+                return new ElementTag(battle.getGatewayCount());
+            });
+
+            // <--[tag]
+            // @attribute <WorldTag.healing_crystals>
+            // @returns ListTag(EntityTag)
+            // @description
+            // Returns a ListTag of the healing end crystals located on top of the obsidian towers.
+            // Only works if the world is an end world.
+            // -->
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java]: These tags are missing the Paper plugin requirement/group meta

File: paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java [hunk 18]
```diff
@@ -23,6 +29,82 @@ public static void register() {
...
+                    return null;
+                }
+                return new ListTag(battle.getHealingCrystals(), EntityTag::new);
+            });
+
+            // <--[tag]
+            // @attribute <WorldTag.respawn_crystals>
+            // @returns ListTag(EntityTag)
+            // @description
+            // Returns a ListTag of the respawn crystals located at the end exit portal.
```
Comment (tal5) [inline paper/src/main/java/com/denizenscript/denizen/paper/properties/PaperWorldExtensions.java]: Nitpick, but I'd probably mention end crystals here (like in the `WorldTag.healing_crystals` meta)

File: plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java [hunk 19]
```diff
@@ -1006,6 +1008,97 @@ else if (time >= 12500) {
...
+
+            // <--[mechanism]
+            // @object WorldTag
+            // @name first_dragon_killed
+            // @input ElementTag(Boolean)
+            // @description
+            // Set whether the first ender dragon was killed already.
+            // Toggling this value won't really affect anything in the end world, but may be useful when creating custom end worlds.
+            // Only works if the world is an end world.
+            // -->
```
Comment (tal5) [inline plugin/src/main/java/com/denizenscript/denizen/objects/WorldTag.java]: Should have a tag link to `first_dragon_killed` (and vice versa)

**Review by mcmonkey4eva — APPROVED** [review]

**Review by tal5 — APPROVED** [review]

**Pull request merged**